nixos/modules/services/matrix/mautrix-meta.nix

```diff
@@ -0,0 +1,489 @@
+        description = "Mautrix-Meta bridge user";
+      };
+
+      users.groups.mautrix-meta = {};
+
+      systemd.services = lib.attrsets.concatMapAttrs (name: cfg: {
+        matrix-synapse = lib.mkIf (config.services.matrix-synapse.enable) {
+          wants = [ cfg.registrationServiceUnit ];
+          after = [ cfg.registrationServiceUnit ];
+          serviceConfig = {
+            # Allow synapse to read the registration file
+            ReadWritePaths = [ (fullDataDir cfg) ];
+          };
+        };
+
+        "${metaName name}-registration" = {
+          description = "Mautrix-Meta registration generation service - ${metaName name}";
+
+          script = ''
+            # substitute the settings file by environment variables
+            # in this case read from EnvironmentFile
+            test -f '${settingsFile cfg}' && rm -f '${settingsFile cfg}'
+            old_umask=$(umask)
+            umask 0177
+            ${lib.getExe pkgs.envsubst} \
+              -o '${settingsFile cfg}' \
+              -i '${settingsFileUnsubstituted cfg}'
+
+            umask 0127
+
+            has_tokens="$(${lib.getExe pkgs.yq} '.appservice | has("as_token") and has("hs_token")' '${settingsFileUnsubstituted cfg}')"
+            generated=0
+
+            if [[ (! -f '${cfg.registrationFile}') ]]; then
+              echo "Generating registration file"
+              ${lib.getExe upperCfg.package} \
+                --generate-registration \
+                --config='${settingsFile cfg}' \
+                --registration='${cfg.registrationFile}'
+              generated=1
+            fi
+
+            umask 0177
+
+            # Make sure correct tokens are in the registration file
+            if [[ "$has_tokens" == "true" ]]; then
+              echo "Substituting as_token, hs_token inside the registration file"
+              ${lib.getExe pkgs.yq} -s '.[1].as_token = .[0].appservice.as_token
+                | .[1].hs_token = .[0].appservice.hs_token
+                | .[1]' '${settingsFileUnsubstituted cfg}' '${cfg.registrationFile}' \
+                > '${cfg.registrationFile}.tmp'
+              mv '${cfg.registrationFile}.tmp' '${cfg.registrationFile}'
+            fi
+
+            # Generation overrides as_token, hs_token.
+            if [[ "$generated" == "1" || "$has_tokens" == "false" ]]; then
+              echo "Substituting as_token, hs_token inside the config file"
+              ${lib.getExe pkgs.yq} -s '.[0].appservice.as_token = .[1].as_token
+                | .[0].appservice.hs_token = .[1].hs_token
+                | .[0]' '${settingsFile cfg}' '${cfg.registrationFile}' \
+                > '${settingsFile cfg}.tmp'
+              mv '${settingsFile cfg}.tmp' '${settingsFile cfg}'
+            fi
+
+            umask $old_umask
+          ''; #+ lib.optionalString config.services.matrix-synapse.enable ''
+          #   chown mautrix-meta:matrix-synapse '${cfg.registrationFile}'
+          # '';
```

Comment:
The best way to do this is probably to create a separate group that both services can be added to. Then the file can be owned by mautrix-meta but readable by the group.

Reply:
It could stay readable by mautrix-meta-X since they are the user owner, no? So it would be enough to change group owner and give them read permissions. The problem I was facing that the chown was giving no permissions and I couldn't figure out why... that's why commented it out. Any idea? I suppose it cannot be because mautrix-meta-X is not member of that group, can it?

Reply:
Yes, `mautrix-meta-X`. I was just being lazy.

> it would be enough to change group owner

I guess it would but I tend to avoid this because you can only have one group owner. If you create a specific group for this data you can assign any number of readers. But I guess multiple readers is probably not necessary for this data.

> Any idea? I suppose it cannot be because mautrix-meta-X is not member of that group, can it?

Yeah maybe. I always struggle to remember how changing ownership works. Does `chown` log an error?

Reply:
Yes it does, but only permission denied / operation not permitted. I can reproduce it and copy it again, I looked into this yesterday I think, so don't have the exact error on hand. But there really wasn't anything else than operation not permitted :( That made me think if maybe the systemd service cannot chown files, but I was not able to figure out why that would be.

Reply:
Don't worry about this, seems it was really just because mautrix-meta was not member of that group. I've created mautrix-meta-registration group, and now chown is possible.

Reply:
I've created `mautrix-meta-registration` group, and assigned that to both the mautrix-meta users, and matrix-synapse user. It seems to be working with synapse finally!

Reply:
So there was actually a bug, since the group was created only if synapse is active, but it was used every time. I fixed this by including the group every time, even when synapse is deactivated. And synapse user is included in the service only if synapse is enabled.

Reply:
That makes sense. Even if the user isn't registering synapse they will likely want to access this file by some service. So always having the group sounds good.